github/VisualStudio
# Ensure that MEF is initialized using an old style non-async package

This workaround causes the `Scanning new and updated MEF components...` dialog to appear. Without this the `GitHub` pane will be blamed for degrading startup performance.

### What this PR does

It provokes MEF to initialize itself using an old style non-async package. This will happen before any tool windows are activated in the `GitSccProvider` auto-load context.

All the `InitializeMefPackage` package does is call `GetService(typeof(SComponentModel))` when it initializes. Visual Studio knows how to handle synchronous initialization of MEF and pops up the `Scanning new and updated MEF components...` dialog.

The `InitializeMefPackage` package will only cause the `GitHub.VisualStudio` assembly to load, and uses an auto-load context we already use in `GitContextPackage`. It shouldn't degrade performance beyond the initial MEF initialization hit immediately after install.

### How to test

- Open Visual Studio 2017
- Ensure the `GitHub` pane is visible
- Click on `Tools > Clear MEF Component Cache...`
- Allow Visual Studio to restart
- If the fix is working you will see:
![image](https://user-images.githubusercontent.com/11719160/37793123-a627f0ba-2e05-11e8-97f3-beb58367761a.png)
- If not would would see:
![image](https://user-images.githubusercontent.com/11719160/37793161-b9d689b4-2e05-11e8-886d-3510a2a096ae.png)

*NOTE: Visual Studio shows this banner based on an average startup time. You might need to do this a few times to being the average startup time back down. 😢 It should then consistently not appear.*

### Notes

I tried more sophisticated approaches but they didn't work nearly as well (or at all). For example:
- [wip] Allow MEF to refresh its cache on a background thread: #1554

When MEF is refreshing on a background thread, no `Scanning new and updated MEF components...` dialog appears. There were other issues that I wasn't able to resolve.

We could probably resolve this issue by converting the `GitHub` pane into an async initializing tool window. Unfortunately this is only supported on recent versions of Visual Studio 2017. We would need to support two code paths and deal with separate Visual Studio 2017 assemblies.

### Related

- Measuring extension impact in startup
https://docs.microsoft.com/en-us/visualstudio/extensibility/how-to-diagnose-extension-performance

Fixes #1550
